File: cgi-bin/post-test.py
#!/usr/bin/env python3

import os
import sys
import urllib.parse

print("Content-type: text/html\n")

# Read raw POST data from stdin
post_data = ""
if "CONTENT_LENGTH" in os.environ:
    content_length = int(os.environ["CONTENT_LENGTH"])
    post_data = sys.stdin.read(content_length)

# Parse the POST data
form_data = urllib.parse.parse_qs(post_data)
name = form_data.get("name", [""])[0].strip()

# Debugging output (optional)
with open("/tmp/cgi_debug.log", "a") as log:
    log.write(f"Raw POST Data: {post_data}\nParsed Data: {form_data}\n")

# Generate response HTML
html_content = f"""<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>POST Request Test</title>
</head>
<body>
    <h1>Test POST Request</h1>

    <form action="/cgi-bin/post-test.py" method="post">
        <label for="name">Enter your name:</label>
        <input type="text" id="name" name="name" required>
        <button type="submit">Send POST Request</button>
    </form>

    <hr>

    <h3>Received Data:</h3>
    <p>{f"Hello, {name}!" if name else "No data received yet."}</p>

    <!-- Debug Info -->
    <pre>POST Data: {post_data}</pre>
</body>
</html>
"""

print(html_content)

File: cgi-bin/post-test.js
#!/usr/bin/env node

const querystring = require("querystring");

const contentLength = process.env.CONTENT_LENGTH
  ? parseInt(process.env.CONTENT_LENGTH, 10)
  : 0;
let body = "";

if (contentLength > 0) {
  process.stdin.on("data", (chunk) => {
    body += chunk;
  });

  process.stdin.on("end", () => {
    handleRequest(body);
  });
} else {
  handleRequest(""); // Handle cases where there's no POST data
}

function handleRequest(body) {
  const postData = querystring.parse(body);
  const name = postData.name
    ? decodeURIComponent(postData.name.replace(/\+/g, " "))
    : "No data received yet.";

  // Debugging output (optional)
  require("fs").appendFileSync(
    "/tmp/cgi_debug.log",
    `Raw POST Data: ${body}\nParsed Name: ${name}\n`,
  );

  // Generate response
  console.log("Content-type: text/html\n");
  console.log(`<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>POST Request Test</title>
</head>
<body>
    <h1>Test POST Request</h1>
    <form action="/cgi-bin/post-test.js" method="post">
        <label for="name">Enter your name:</label>
        <input type="text" id="name" name="name" required>
        <button type="submit">Send POST Request</button>
    </form>
    <hr>
    <h3>Received Data:</h3>
    <p>${name}</p>
</body>
</html>`);
}
